File: packages/typescript/src/services/research.ts
import { BaseService } from './base.js';

export interface SearchResult {
  results: Array<{
    title: string;
    url: string;
    snippet: string;
  }>;
}

export interface ResearchResult {
  answer: string;
  sources?: Array<{ title: string; url: string }>;
}

export class ResearchService extends BaseService {
  async search(query: string, options?: { maxResults?: number }): Promise<SearchResult> {
    const params = new URLSearchParams({ query });
    if (options?.maxResults != null) params.set('max_results', String(options.maxResults));
    return this.request<SearchResult>('POST', `/v1/research/search?${params}`);
  }

  async research(query: string): Promise<ResearchResult> {
    return this.request('POST', '/v1/research', { body: { query } });
  }

  async deep(query: string): Promise<ResearchResult> {
    return this.request('POST', '/v1/research/deep', { body: { query } });
  }
}

File: packages/python/scalix/services/research.py
"""Research and web search service."""

from __future__ import annotations

from typing import Any

from scalix.services.base import BaseService


class ResearchService(BaseService):
    async def search(
        self, query: str, *, max_results: int | None = None
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"query": query}
        if max_results:
            params["max_results"] = max_results
        return await self._request("POST", "/v1/research/search", params=params)

    async def research(self, query: str) -> dict[str, Any]:
        return await self._request("POST", "/v1/research", json={"query": query})

    async def deep(self, query: str) -> dict[str, Any]:
        return await self._request("POST", "/v1/research/deep", json={"query": query})
